infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,85 +101,127 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
      * @return global address
      */
     public Long getIfPresent(ConflictTxStream conflictKey) {
-        return conflictCache.getIfPresent(conflictKey);
+        return cacheConflictKeys.get(conflictKey);
     }
 
     /**
-     * Invalidate all records up to a trim mark.
-     *
-     * @param trimMark trim mark
+     * The first address in the priority queue.
      */
-    public void invalidateUpTo(long trimMark) {
-        log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
-
-        AtomicLong entries = new AtomicLong();
+    public long firstAddress() {
+        if (cacheEntries.isEmpty()) {
+            return Address.NOT_FOUND;
+        }
+        return cacheEntries.peek().txVersion;
+    }
 
-        conflictCache.asMap().forEach((key, txVersion) -> {
-            if (txVersion == null || txVersion >= trimMark) {
-                return;
-            }
+    /**
+     * Invalidate the records with the minAddress. It could be one or multiple records
+     * @return the number of entries has been invalidated and removed from the cache.
+     */
+    private int invalidateFirst() {
+        ConflictTxStream firstEntry = cacheEntries.peek();
+        if (firstEntry == null) {
+            return 0;
+        }
 
-            conflictCache.invalidate(key);
-            entries.incrementAndGet();
-        });
+        int i = 0;
+        while (!cacheEntries.isEmpty() && cacheEntries.peek().txVersion == firstEntry.txVersion) {
```

Comment:
Looks like you can use the `firstAddress()` method you added to simplify this logic: `while (firstAddress() == firstEntry.txVersion)`